storage/database/src/tables.rs

```diff
@@ -0,0 +1,427 @@
+            // ```rust
+            // (BLOCKINFOSV1S, BLOCKINFOSV2S, BLOCKINFOSV3S, [...])
+            // ```
+            // which is just a tuple of the generics defined above.
+            for ($([<$table:upper>]),*)
+        where
+            // This expands to a where bound that asserts each element
+            // in the tuple implements some database table type.
+            //
+            // Concretely, this expands to something like:
+            // ```rust
+            // BLOCKINFOSV1S: DatabaseRo<BlockInfoV1s> + DatabaseIter<BlockInfoV1s>,
+            // BLOCKINFOSV2S: DatabaseRo<BlockInfoV2s> + DatabaseIter<BlockInfoV2s>,
+            // [...]
+            // ```
+            $(
+                [<$table:upper>]: $crate::DatabaseRo<$table>,
+            )*
+        {
+            $(
+                // The function name of the accessor function is
+                // the table type in `snake_case`, e.g., `block_info_v1s()`.
+                #[inline]
+                fn [<$table:snake>](&self) -> &impl $crate::DatabaseRo<$table> {
+                    // The index of the database table in
+                    // the tuple implements the table trait.
+                    &self.$index
+                }
+            )*
+
+            fn all_tables_empty(&self) -> Result<bool, $crate::RuntimeError> {
+                $(
+                     if !$crate::DatabaseRo::is_empty(&self.$index)? {
+                        return Ok(false);
+                     }
+                )*
+                Ok(true)
+            }
+        }
+
+        // This is the same as the above
+        // `Tables`, but for `TablesIter`.
+        impl<$([<$table:upper>]),*> TablesIter
+            for ($([<$table:upper>]),*)
+        where
+            $(
+                [<$table:upper>]: $crate::DatabaseRo<$table> + $crate::DatabaseIter<$table>,
+            )*
+        {
+            $(
+                // The function name of the accessor function is
+                // the table type in `snake_case` + `_iter`, e.g., `block_info_v1s_iter()`.
+                #[inline]
+                fn [<$table:snake _iter>](&self) -> &(impl $crate::DatabaseRo<$table> + $crate::DatabaseIter<$table>) {
+                    &self.$index
+                }
+            )*
+        }
+
+        // This is the same as the above
+        // `Tables`, but for `TablesMut`.
+        impl<$([<$table:upper>]),*> TablesMut
+            for ($([<$table:upper>]),*)
+        where
+            $(
+                [<$table:upper>]: $crate::DatabaseRw<$table>,
+            )*
+        {
+            $(
+                // The function name of the mutable accessor function is
+                // the table type in `snake_case` + `_mut`, e.g., `block_info_v1s_mut()`.
+                #[inline]
+                fn [<$table:snake _mut>](&mut self) -> &mut impl $crate::DatabaseRw<$table> {
+                    &mut self.$index
+                }
+            )*
+        }
```

Comment:
### Macro internals 2/3
Create `trait {Tables,TablesIter,TablesMut}` specific to these tables.